gc/base/standard/ConcurrentGC.cpp

```diff
@@ -1427,7 +1426,59 @@ MM_ConcurrentGC::tuneToHeap(MM_EnvironmentBase *env)
+			 * should be 0-ed or 1-ed).
+			 *
+			 * Hold _initWorkMonitor to prevent any thread from starting/joining/leaving init work,
+			 * since we are about to change the _initRange table (that initializer threads consume) */
+			omrthread_monitor_enter(_initWorkMonitor);
+			/* Are there threads that already started consuming _initRange table? */
+			while (_initializers > 0) {
+				/* Yes, they are. We will wait for them to finish. _initWorkCompleteMonitor monitor is used to wait on,
+				 * the same one initializers use to notify themselves the init work is complete.
+				 *
+				 * We will also set _determineInitWorkPending flag, to let initializers now they need to notify us,
+				 * in case the last active initializers leave init work (by paying enough tax) without actually finishing init work (or even without being pre-empted
+				 * by exclusive access request). This condition (no active initializers and init work not complete) is temporary and new initializers will typically show up
+				 * to complete init work. But not always, because exclusive access might have been requested during this temporary condition (rather improbable, but possible).
+				 * If that occurs, there would be a danger of this thread staying blocked for ever, and holding VM access. To prevent it, _determineInitWorkPending
+				 * will help initializers realize there is one more party to be notifed of this temporary condition. Once notified we will try to sneak in and reconfigure _initRange.
+				 * If new initializers do show up, no problem, we will give up again our attempt, and wait till they really finish. If they do not show up (because they released their VM access),
+				 * we will complete _initRanges reconfiguration and proceed.
+				 */
+				_determineInitWorkPending = true;
+				omrthread_monitor_enter(_initWorkCompleteMonitor);
+
+				/* While waiting to be notified initialization is complete,
+				 * we don't want to prevent other threads to join the ongoing initialization,
+				 * so we will release _initWorkMonitor (that controls initializes joining) just before blocking. */
+				omrthread_monitor_exit(_initWorkMonitor);
+
+				omrthread_monitor_wait(_initWorkCompleteMonitor);
+
+				/* Ongoing initialization just (likely) completed. Re-acquire _initWorkMonitor to prevent any other thread to join (in case this was just a temporary condition
+				 *  where work was not complete), because we are about to examine if we can complete our work. We will not try to distinguish the case between cases initialization
+				 *  being still in progress, or being completed. Either way, we will do our work. It would not be correct to wait them to finish, because they might have given up
+				 *  VM access and might not come back. Reconfiguring table that has been already used to do some init work, will cause the init work to be redone,
+				 *  but that's a reasonable price to pay for such a rare scenario. */
+				omrthread_monitor_enter(_initWorkMonitor);
```

Comment:
I believe this could cause a deadlock.  There are locations including the new code above that have the following pattern with monitors:
```
enter(_initWorkMonitor)
enter(_initWorkCompleteMonitor)
```

This new code will be holding `_initWorkCompleteMonitor` and then trying to enter `_initWorkMonitor` which will make the enter order:
```
enter(_initWorkCompleteMonitor)
enter(_initWorkMonitor)
```

If threads are entering the monitors in different orders I believe we could deadlock

Reply:
Valid concern - but things are not so simple. Let's see....

This order 
enter(_initWorkCompleteMonitor)
enter(_initWorkMonitor)
occurs only at one point, in the new code, when the heap expanding thread, waiting for Mark Map initializer threads to complete the work, is notified that there are no more active initializers (0 ==_initializers). When this thread tries to acquire _initWorkMonitor again (already holding _initWorkCompleteMonitor), no other thread should really attempt to acquire _initWorkCompleteMonitor (while already holding _initWorkMonitor).

There are 3 spots that the later could potentially occur:
1) another thread trying to expand heap (entering this new code sequence). It is not possible, because it's prevented (even in standard Scavenger) by Subspace level lock that serialize such threads.

2) end of doConcurrentInitialization when new initalizers threads try to leave and potentially notify any other blocking thread (either initalizers ones or heap expanding ones). It should not be possible because there no active initializers threads (0 ==_initializers) nor new one should be able to join (see 3)

3) beginning of doConcurrentInitialization, when new initalizers threads try to join.
A new thread trying to join will quickly realize that it should not join, for the same reasons one of initalizers threads recently notified the heap expanding threads there are no more active initializers: init work complete or exclusive access has been requested. If any of two checks are true, the thread will release _initWorkMonitor and leave without attempting to acquire _initWorkCompleteMonitor. 
Things are, however, even more complicated with _determineInitWorkPending flag, because it is one of the conditions we notify heap expanding flag, but it is not the reason to prevent new initializes to join (because we want them to help and finish initialization as fast as possible). I'll think about this more, if could cause a deadlock or not... (I'll also think about possibly simplify things by making heap expanding thread not blocking at all, but just postpone the init table update for a later, safer point).